drivers/Watchdog.cpp

```diff
@@ -18,48 +18,103 @@
 
 #include "Watchdog.h"
 
-namespace mbed {
+namespace mbed
+{
+
+Watchdog *Watchdog::first;
 
-watchdog_status_t Watchdog::start(const uint32_t timeout)
+bool Watchdog::add_to_list()
 {
-    if (timeout == 0) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
+    Watchdog *node;
+    node = first;
+    if(first == NULL) {
+        first = this;
+        this->next=NULL;
+    } else {
+        while(node->next != NULL) {
+            node = node->next;
+            if(node == this) {
+                return false;
+            }
+        }
+        this->next = NULL;
+        node->next = this;
     }
+    return true;
+}
 
-    if (timeout > max_timeout()) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
+void Watchdog::start()
+{
+    bool msts = false;
+    MBED_ASSERT(!_is_initialized);
+    core_util_critical_section_enter();
+    msts = add_to_list();
+    if(msts) {
+        _is_initialized = true;
     }
-
-    watchdog_config_t config;
-    config.timeout_ms = timeout;
-
-    return hal_watchdog_init(&config);
+    core_util_critical_section_exit();
 }
 
 
 void Watchdog::kick()
 {
-    hal_watchdog_kick();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    _current_count = 0;
+    core_util_critical_section_exit();
 }
 
-
-watchdog_status_t Watchdog::stop()
+void Watchdog::stop()
 {
-    return hal_watchdog_stop();
+    bool msts = false;
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    msts = remove_from_list();
+    if(msts) {
+        _is_initialized = false;
```

Comment:
Should `_is_initialized` be set to false unconditionally? `remove_from_list` returning false means that this object wasn't in the list in the first place right?